### PR TITLE
Complete update to layout management

### PR DESCRIPTION
**Pull request by Damillora**
This commit completes the move to layout management started by pull
request #15

Related to https://github.com/HaikuArchives/FtpPositive/pull/15

### CODE REVIEW

File: src/BookmarkWindow.cpp [hunk 1]
```diff
@@ -24,26 +24,23 @@ static const char* kCancel = B_TRANSLATE("Cancel");
...
 	
-	fBookmarkName = new TOAEdit(BRect(lx, 20, rx, 38), "BookmarkName", B_TRANSLATE("Name:"), B_TRANSLATE("New bookmark"), new BMessage(EDIT_CHANGED));
-	fHostAddress = new TOAEdit(BRect(lx, 40, rx, 58), "HostAddress", B_TRANSLATE("Address:"), "", new BMessage(EDIT_CHANGED));
-	fPort = new TOAEdit(BRect(lx, 60, rx, 78), "Port", B_TRANSLATE("Port:"), "21", new BMessage(EDIT_CHANGED));
-	fUsername = new TOAEdit(BRect(lx, 80, rx, 98), "Username", B_TRANSLATE("User name:"), "", new BMessage(EDIT_CHANGED));
-	fPassword = new TOAEdit(BRect(lx, 100, rx, 118), "Password", B_TRANSLATE("Password:"), "", new BMessage(EDIT_CHANGED));
-	fEncoder = new TOAEdit(BRect(lx, 120, rx, 138), "Encoder", B_TRANSLATE("Encoder:"), "", NULL);
-	fRemotePath = new TOAEdit(BRect(lx, 140, rx, 158), "RemotePath", B_TRANSLATE("Remote path:"), "", NULL);
-	fLocalPath = new TOAEdit(BRect(lx, 160, rx, 178), "LocalPath", B_TRANSLATE("Local path:"), TFtpPositive::GetDefaultLocalDir().String(), NULL);
+	fBookmarkName = new TOAEdit(BRect(0,0,1,1),"BookmarkName", B_TRANSLATE("Name:"), B_TRANSLATE("New bookmark"), new BMessage(EDIT_CHANGED));
```
Comment (humdingerb) [inline src/BookmarkWindow.cpp]: When using layout managment, you need to use the layout-aware constructors, i.e. without the BRect for the frame etc. See https://www.haiku-os.org/docs/api/classBTextControl.html
You'll have to adjust OAEdit.cpp

File: src/BookmarkWindow.cpp [hunk 2]
```diff
@@ -56,15 +53,15 @@ TBookmarkWindow::TBookmarkWindow(float x, float y, const char *title, const char
...
 	
-	fSaveButton = new BButton(BRect(150, 215, 210, 240), "Save", B_TRANSLATE("Save"), new BMessage(SAVE_CLICKED));
+	fSaveButton = new BButton("Save", B_TRANSLATE("Save"), new BMessage(SAVE_CLICKED));
 	fSaveButton->SetEnabled(false);
 	
-	BButton *cancelButton = new BButton(BRect(220, 215, 280, 240), "Cancel", kCancel, new BMessage(B_QUIT_REQUESTED));
+	BButton *cancelButton = new BButton("Cancel", kCancel, new BMessage(B_QUIT_REQUESTED));
 	
-	BLayoutBuilder::Group<>(bgview,B_VERTICAL,2)
+	BLayoutBuilder::Group<>(this,B_VERTICAL,2)
```
Comment (humdingerb) [inline src/BookmarkWindow.cpp]: Add a .SetInsets(B_USE_WINDOW_INSETS) for some padding.
It'd actually look nicer by using a grid layout. Have a look at http://xref.plausible.coop/source/xref/haiku/src/preferences/network/InterfaceAddressView.cpp#103 as an example

File: src/ChmodWindow.cpp [hunk 3]
```diff
@@ -53,7 +50,8 @@ TChmodWindow::TChmodWindow(float ix, float iy, const char *title)
 		.AddGrid(10,10)
 			.Add(fOKButton,0,0)
 			.Add(fCancelButton,1,0)
-		.End();
+		.End()
+		.View()->SetViewColor(217,217,217);
```
Comment (humdingerb) [inline src/ChmodWindow.cpp]: Better use ui_color(B_PANEL_BACKGROUND_COLOR). Needs ControlLook.h I think.
Maybe have a look where else colors are hard-coded.

File: src/ProgressWindow.cpp [hunk 4]
```diff
@@ -18,7 +18,7 @@ extern const char* kAppName;
 // TProgressView
 
 TProgressView::TProgressView(BRect frame, const char *name)
-	:	BView(frame, name, B_FOLLOW_LEFT | B_FOLLOW_TOP, B_PULSE_NEEDED)
+	:	BView(frame, name, B_FOLLOW_LEFT | B_FOLLOW_TOP, B_PULSE_NEEDED | B_SUPPORTS_LAYOUT)
```
Comment (humdingerb) [inline src/ProgressWindow.cpp]: This should also use the layout-aware c'tor without the BRect and B_FOLLOW_* flags

**Comment by Damillora** [comment]
Fixed all where colors are hard coded, and layout aware c'tors

The BookmarkWindow changes are already commited but somehow not pushed before, now it should be there.

File: src/BookmarkWindow.cpp [hunk 5]
```diff
@@ -56,28 +54,38 @@ TBookmarkWindow::TBookmarkWindow(float x, float y, const char *title, const char
...
+		.Add(fUsername->CreateTextViewLayoutItem(),1,3)
+		.Add(fPassword->CreateLabelLayoutItem(),0,4)
+		.Add(fPassword->CreateTextViewLayoutItem(),1,4)
+		.Add(fEncoder->CreateLabelLayoutItem(),0,5)
+		.Add(fEncoder->CreateTextViewLayoutItem(),1,5)
+		.Add(fRemotePath->CreateLabelLayoutItem(),0,6)
+		.Add(fRemotePath->CreateTextViewLayoutItem(),1,6)
+		.Add(fLocalPath->CreateLabelLayoutItem(),0,7)
+		.Add(fLocalPath->CreateTextViewLayoutItem(),1,7)
+		.AddGroup(B_HORIZONTAL,10,0,8,2)
```
Comment (humdingerb) [inline src/BookmarkWindow.cpp]: Could add a .AddStrut(B_USE_BIG_SPACING) in front of the buttons.

**Comment by Damillora** [comment]
BPictureButton ued in TSimplePictureButton doesn't have a layout c'tor (consulted the Haiku API documentation)

File: src/BookmarkWindow.cpp (hunk 5 above)
Comment (Damillora) [reply src/BookmarkWindow.cpp]: Changes done

**Comment by Damillora** [comment]
I also made the navigation buttons use BToolBar, since BPictureButton should not be used anymore :)

**Comment by Damillora** [comment]
The usage of BToolBar make the TSimplePictureButton unnecessary, and only a single helper function remains. Trying to figure out where to move the helper function...

File: src/BookmarkWindow.cpp (hunk 5 above)
Comment (humdingerb) [reply src/BookmarkWindow.cpp]: The strut should be before the horizontal group.

File: src/ChmodWindow.cpp [hunk 6]
```diff
@@ -53,7 +51,9 @@ TChmodWindow::TChmodWindow(float ix, float iy, const char *title)
 		.AddGrid(10,10)
```
Comment (humdingerb) [inline src/ChmodWindow.cpp]: Maybe add a strut here before the button group, too.

File: src/FTPWindow.cpp [hunk 7]
```diff
@@ -32,8 +33,8 @@ static const rgb_color kBackgroundColor = {0xff,0xff,0xff,0xff};
 
 // ----------------------------------- TLogView -------------------------------------
 
-TLogView::TLogView(BRect rect, const char *name, BRect textRect)
-	:	BTextView(rect, name, textRect, B_FOLLOW_ALL_SIDES, B_WILL_DRAW)
+TLogView::TLogView( const char *name)
```
Comment (humdingerb) [inline src/FTPWindow.cpp]: stray space

File: src/FTPWindow.h [hunk 8]
```diff
@@ -78,24 +78,20 @@ class TFTPWindow : public BWindow
 private:
 	TFtpLooper *fFtpLooper;
 	TConfigFile *fBookmarkConfig;
-	
+
```
Comment (humdingerb) [inline src/FTPWindow.h]: stray tab

File: src/RenameWindow.cpp [hunk 9]
```diff
@@ -14,25 +14,24 @@ TRenameWindow::TRenameWindow(float x, float y,
...
-	fOKButton = new BButton(BRect(0,0,1,1), "", B_TRANSLATE("OK"), new BMessage(OK_CLICKED));
-	fCancelButton = new BButton(BRect(0,0,1,1), "", B_TRANSLATE("Cancel"), new BMessage(B_QUIT_REQUESTED));
+	fOKButton = new BButton("", B_TRANSLATE("OK"), new BMessage(OK_CLICKED));
+	fCancelButton = new BButton("", B_TRANSLATE("Cancel"), new BMessage(B_QUIT_REQUESTED));
 
-	BLayoutBuilder::Group<>(bgView,B_VERTICAL,10)
+	BLayoutBuilder::Group<>(this,B_VERTICAL,10)
 		.SetInsets(10)
 		.Add(fTextControl)
 		.AddGrid(10,10)
```
Comment (humdingerb) [inline src/RenameWindow.cpp]: Maybe another strut before the buttons group. which could as well be a horizontal group instead of a grid.

**Comment by Damillora** [comment]
Requested changes done, along with some layout fixes I discovered again.

**Comment by humdingerb** [comment]
One tiny thing, and we merge. Promise...
Instead of hard coding the spacing around stuff, can you use the constants from http://cgit.haiku-os.org/haiku/tree/headers/os/interface/InterfaceDefs.h#n231
e.g. B_USE_ITEM_SPACING.
Some things, maybe the permissions checkboxes may look nicer with B_USE_HALF_ITEM_SPACING or B_USE_SMALL_SPACING.

**Comment by Damillora** [comment]
Noted, and the constants are now used in the layout.

**Comment by humdingerb** [comment]
Thank you!

**Comment by scottmc** [comment]
Perhaps you can update FtpPositiveActive.png with a screenshot of what it looks like now with the recent changes?  Be sure to use the option to capture current window with tab.

**Comment by Damillora** [comment]
Created https://github.com/HaikuArchives/FtpPositive/pull/20 to update screenshot.

**Pull request merged**